plugins/inputs/prometheus/kubernetes.go

```diff
@@ -98,15 +99,44 @@ func (p *Prometheus) watch(ctx context.Context, client *k8s.Client) error {
 
 			switch eventType {
 			case k8s.EventAdded:
-				registerPod(pod, p)
-			case k8s.EventDeleted:
-				unregisterPod(pod, p)
+				if podReady(pod.Status.GetContainerStatuses()) {
+					registerPod(pod, p)
+				}
 			case k8s.EventModified:
+				if pod.Metadata.GetDeletionTimestamp() != nil && podConditionReady(pod.Status.GetConditions()) {
+					unregisterPod(pod, p)
+				} else if podReady(pod.Status.GetContainerStatuses()) && podConditionReady(pod.Status.GetConditions()) {
+					registerPod(pod, p)
+				}
 			}
 		}
 	}
 }
 
+func podReady(statuss []*corev1.ContainerStatus) bool {
+	if len(statuss) == 0 {
+		return false
+	}
+	for _, cs := range statuss {
+		if !cs.GetReady() {
+			return false
+		}
+	}
+	return true
+}
+
+func podConditionReady(statuss []*corev1.PodCondition) bool {
+	if len(statuss) < 3 {
+		return false
+	}
+	for _, pc := range statuss {
+		if pc.GetType() == "Ready" && pc.GetStatus() == "False" {
```

Comment:
This looks brittle, relying on the number of conditions and finding a Ready state... Is there only one condition with GetStatus() == "True" at a time?

Reply:
Any and all can have different or the same status values.

Reply:
Couldn't think of a great way to avoid checking the length, so I just went with the less brittle way to check.

If `len(pod.Status.GetContainerStatuses()) == 0` then there are no containers in the pod and it is not ready, thus `false`. If any containers are not "ready", the pod appears to not be fully ready either.